hathor/utils/yaml.py

```diff
@@ -11,13 +11,57 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
-
-from typing import Any, Dict
+import os
+from pathlib import Path
+from typing import Any, Union
 
 import yaml
 
+from hathor.utils.dict import deep_merge
+
+_EXTENDS_KEY = 'extends'
 
-def dict_from(*, filepath: str) -> Dict[str, Any]:
+
+def dict_from_yaml(*, filepath: Union[Path, str]) -> dict[str, Any]:
     """Takes a filepath to a yaml file and returns a dictionary with its contents."""
+    if not os.path.isfile(filepath):
+        raise ValueError(f"'{filepath}' is not a file")
+
     with open(filepath, 'r') as file:
-        return yaml.safe_load(file)
+        contents = yaml.safe_load(file)
+
+        if contents is None:
+            return {}
+
+        if not isinstance(contents, dict):
+            raise ValueError(f"'{filepath}' cannot be parsed as a dictionary")
+
+        return contents
+
+
+def dict_from_extended_yaml(*, filepath: Union[Path, str]) -> dict[str, Any]:
+    """
+    Takes a filepath to a yaml file and returns a dictionary with its contents.
+    Supports extending another yaml file via the 'extends' key in the file.
+
+    Note: the 'extends' key is reserved and will not be present in the returned dictionary.
+    To opt-out of the extension feature, use dict_from_yaml().
+    """
+    extension_dict = dict_from_yaml(filepath=filepath)
+    base_file = extension_dict.pop(_EXTENDS_KEY, None)
+
+    if not base_file:
+        return extension_dict
+
+    root_path = Path(filepath).parent
+    base_filepath = root_path / str(base_file)
+
+    if not os.path.isfile(base_filepath):
+        raise ValueError(f"'{base_filepath}' is not a file")
+
+    assert base_filepath.resolve() != Path(filepath).resolve(), 'cannot extend self'
+
+    base_dict = dict_from_yaml(filepath=base_filepath)
```

Comment:
This means you can't extend configs that already use extend, right? Is there anything preventing doing a recursion here so we can support that, other than not detecting A -> B -> A infinite extensions?

I slightly prefer being able to extend config multiple times, although I wouldn't use it. But I'm OK not supporting that, although I think that should be made more explicit somewhere else, or maybe it already is and I missed it.

Reply:
Nope, I missed that. Implemented in https://github.com/HathorNetwork/hathor-core/pull/618/commits/9a41c03e8b5e42fad57aa93363e3267d37d65007